src/ripple/app/tx/impl/SetTrust.cpp

```diff
@@ -128,20 +129,46 @@ SetTrust::preclaim(PreclaimContext const& ctx)
         }
     }
 
+    // This might be nullptr
+    auto const sleDst = ctx.view.read(keylet::account(uDstAccountID));
+
     // If the destination has opted to disallow incoming trustlines
     // then honour that flag
     if (ctx.view.rules().enabled(featureDisallowIncoming))
     {
-        auto const sleDst = ctx.view.read(keylet::account(uDstAccountID));
-
         if (!sleDst)
             return tecNO_DST;
 
-        auto const dstFlags = sleDst->getFlags();
-        if (dstFlags & lsfDisallowIncomingTrustline)
+        if (sleDst->getFlags() & lsfDisallowIncomingTrustline)
             return tecNO_PERMISSION;
     }
 
+    // If destination is AMM and the trustline doesn't exist then only
+    // allow SetTrust if the asset is AMM LP token and AMM is not
+    // in empty state.
+    if (ammEnabled(ctx.view.rules()))
+    {
+        if (!sleDst)
+            return tecNO_DST;
+
+        if (sleDst->isFieldPresent(sfAMMID) &&
+            !ctx.view.read(keylet::line(id, uDstAccountID, currency)))
```

Comment:
What happens if the issuer of one of the two currencies that the AMM holds uses SetTrust to alter the lines created by the AMM? It's just allowed to happen? What if they set no-ripple, freeze, etc?